# Teste Amostral da Linha 010

Código feito por Igor Laltuf, ago/23

---

## Objetivo

Este notebook tem como objetivo realizar o teste amostral da linha 010 com dados da primeira quinzena de janeiro de 2023.

## Etapas
1) Comparar dados do gabarito enviado pelos consórcios com os dados da solução proposta pela DTDI.

2) Verificar se os veículos não identificados na solução operaram nos dias e horários apontados no gabarito, com base em dados de GPS e de conformidade das viagens.

## Conclusão

Ao considerar um intervalo de 10 minutos para mais ou para menos entre os horários de partida, foram encontradas 699 das 706 viagens apresentadas no gabarito (94,75% das viagens do gabarito).

Após verificar as sete viagens não identificadas na solução acima, concluiu-se que:

- Não foram encontrados dados das quatro viagens do veículo A72037 na linha 010, mas foram encontrados sinais de GPS em horários compatíveis com duas das quatro viagens, operando na linha 007.

- No caso do veículo A72085, na viagem do dia 15/01/2023 foram encontrados sinais de GPS para o mesmo veículo, dia e serviço (mas não foram encontrados dados para este dia na tabela de conformidade). Quanto a viagem do dia 04/01/2023, foram identificados sinais de GPS apenas em horário próximo ao término da viagem.

- Para o veículo A72151, foi encontrada uma viagem da linha 010 entre 18:31 e 20:15, horário que se sobrepõe ao horário da viagem apontada no gabarito, entre 19:19 e 20:17.



### Carregar bibliotecas 

In [78]:
import basedosdados as bd
import pandas as pd
import numpy as np
import plotly.express as px

from datetime import timedelta, datetime

# Paths
import os
import sys
from pathlib import Path 

current_path = Path().resolve()

paths = dict()
paths["raw"] = current_path / 'data' / 'raw'
paths["treated"] = current_path / 'data' / 'treated'

for path in paths.values():
    if not os.path.exists(path):
        os.makedirs(path)

# Seta o projeto a ser cobrado
bd.config.billing_project_id = "rj-smtr-dev"

# Seta as configurações a serem usadas
bd.config.project_config_path = "C:/Users/igorl/.basedosdados/"
# bd.config.project_config_path = "D:\\basedosdados\\dev"

### 1. Importar dados do gabarito e da solução

In [79]:
solucao = pd.read_excel('data/raw/amostra_linha_010_q1_2023_tratado.xlsx', sheet_name=0)  
solucao

data  servico sentido id_veiculo    datetime_partida  \
0   2023-01-02       10       I     A72037 2023-01-02 06:07:57   
1   2023-01-02       10       V     A72037 2023-01-02 06:32:00   
2   2023-01-02       10       I     A72085 2023-01-02 06:35:11   
3   2023-01-02       10       I     A72037 2023-01-02 07:00:04   
4   2023-01-02       10       V     A72085 2023-01-02 07:05:29   
..         ...      ...     ...        ...                 ...   
729 2023-01-15       10       V     A72085 2023-01-15 19:18:23   
730 2023-01-15       10       I     A72085 2023-01-15 19:42:45   
731 2023-01-15       10       V     A72085 2023-01-15 20:08:57   
732 2023-01-15       10       I     A72085 2023-01-15 20:32:29   
733 2023-01-15       10       V     A72085 2023-01-15 21:01:50   

       datetime_chegada  extensao  
0   2023-01-02 06:19:58      5505  
1   2023-01-02 06:43:21      4902  
2   2023-01-02 06:47:51      5505  
3   2023-01-02 07:13:44      5505  
4   2023-01-02 07:21:50      4902  
..                  ...       ...  
729 2023-01-15 19:32:43      4902  
730 2023-01-15 19:57:06      5505  
731 2023-01-15 20:20:57      4902  
732 2023-01-15 20:50:30      5505  
733 2023-01-15 21:25:35      4902  

[734 rows x 7 columns]

In [80]:
gabarito = pd.read_excel('data/raw/amostra_linha_010_q1_2023_tratado.xlsx', sheet_name=1) 
gabarito

data  servico sentido id_veiculo    datetime_partida  \
0   2023-01-02       10       I     A72037 2023-01-02 06:07:00   
1   2023-01-02       10       V     A72037 2023-01-02 06:29:00   
2   2023-01-02       10       I     A72085 2023-01-02 06:33:00   
3   2023-01-02       10       I     A72037 2023-01-02 06:58:00   
4   2023-01-02       10       V     A72085 2023-01-02 07:03:00   
..         ...      ...     ...        ...                 ...   
702 2023-01-15       10       I     A72085 2023-01-15 18:51:00   
703 2023-01-15       10       V     A72085 2023-01-15 19:16:00   
704 2023-01-15       10       I     A72085 2023-01-15 19:41:00   
705 2023-01-15       10       V     A72085 2023-01-15 20:06:00   
706 2023-01-15       10       I     A72085 2023-01-15 20:30:00   

       datetime_chegada  extensao  
0   2023-01-02 06:22:00      5505  
1   2023-01-02 06:56:00      4902  
2   2023-01-02 06:52:00      5505  
3   2023-01-02 07:16:00      5505  
4   2023-01-02 07:29:00      4902  
..                  ...       ...  
702 2023-01-15 19:14:00      5505  
703 2023-01-15 19:38:00      4902  
704 2023-01-15 19:59:00      5505  
705 2023-01-15 20:28:00      4902  
706 2023-01-15 20:53:00      5505  

[707 rows x 7 columns]

### 2. Tratar dados do gabarito

Remover os casos em que o mesmo veículo fez duas viagens no mesmo dia e horário.

In [81]:

gabarito = gabarito.sort_values(by=['id_veiculo', 'datetime_partida', 'datetime_chegada'])

# Filtrar linhas com base nas condições de sobreposição
mask = (
    gabarito.groupby('id_veiculo')['datetime_chegada'].shift(1).fillna(pd.Timestamp.min) <= gabarito['datetime_partida']
) & (
    gabarito['datetime_chegada'] <= gabarito.groupby('id_veiculo')['datetime_partida'].shift(-1).fillna(pd.Timestamp.max)
)
gabarito_tratado = gabarito[mask]

gabarito_tratado


data  servico sentido id_veiculo    datetime_partida  \
0   2023-01-02       10       I     A72037 2023-01-02 06:07:00   
1   2023-01-02       10       V     A72037 2023-01-02 06:29:00   
3   2023-01-02       10       I     A72037 2023-01-02 06:58:00   
5   2023-01-02       10       V     A72037 2023-01-02 07:28:00   
8   2023-01-02       10       I     A72037 2023-01-02 08:02:00   
..         ...      ...     ...        ...                 ...   
100 2023-01-03       10       I     A72151 2023-01-03 17:50:00   
102 2023-01-03       10       V     A72151 2023-01-03 18:27:00   
104 2023-01-03       10       I     A72151 2023-01-03 19:19:00   
105 2023-01-03       10       V     A72151 2023-01-03 20:18:00   
106 2023-01-03       10       I     A72151 2023-01-03 21:03:00   

       datetime_chegada  extensao  
0   2023-01-02 06:22:00      5505  
1   2023-01-02 06:56:00      4902  
3   2023-01-02 07:16:00      5505  
5   2023-01-02 07:58:00      4902  
8   2023-01-02 08:24:00      5505  
..                  ...       ...  
100 2023-01-03 18:25:00      5505  
102 2023-01-03 19:13:00      4902  
104 2023-01-03 20:17:00      5505  
105 2023-01-03 21:00:00      4902  
106 2023-01-03 21:20:00      5505  

[705 rows x 7 columns]

In [82]:
# Verificar viagens do mesmo veículo no mesmo dia e horário
sobrepostos = gabarito.merge(gabarito_tratado, how='left', indicator=True)
sobrepostos = sobrepostos[sobrepostos['_merge'] == 'left_only']

sobrepostos


data  servico sentido id_veiculo    datetime_partida  \
393 2023-01-03       10       V     A72085 2023-01-03 21:18:00   
394 2023-01-03       10       I     A72085 2023-01-03 21:20:00   

       datetime_chegada  extensao     _merge  
393 2023-01-03 21:37:00      4902  left_only  
394 2023-01-03 21:36:00      5505  left_only

In [83]:
# Modificar gabarito_tratado para incluir uma das linhas do df sobrepostos
gabarito_tratado = pd.concat([gabarito_tratado, sobrepostos.iloc[[0]]])
gabarito_tratado = gabarito_tratado[['id_veiculo', 'datetime_partida', 'datetime_chegada']]
gabarito_tratado


id_veiculo    datetime_partida    datetime_chegada
0       A72037 2023-01-02 06:07:00 2023-01-02 06:22:00
1       A72037 2023-01-02 06:29:00 2023-01-02 06:56:00
3       A72037 2023-01-02 06:58:00 2023-01-02 07:16:00
5       A72037 2023-01-02 07:28:00 2023-01-02 07:58:00
8       A72037 2023-01-02 08:02:00 2023-01-02 08:24:00
..         ...                 ...                 ...
102     A72151 2023-01-03 18:27:00 2023-01-03 19:13:00
104     A72151 2023-01-03 19:19:00 2023-01-03 20:17:00
105     A72151 2023-01-03 20:18:00 2023-01-03 21:00:00
106     A72151 2023-01-03 21:03:00 2023-01-03 21:20:00
393     A72085 2023-01-03 21:18:00 2023-01-03 21:37:00

[706 rows x 3 columns]

Na solução encontramos 734 viagens realizadas, contra 706 viagens realizadas no gabarito após o tratamento.

### 3. Comparar as viagens do gabarito com as viagens da solução

Para realizar a comparação entre os dados, foi considerado um desvio de 10 minutos para mais ou para menos no horário de partida.

In [84]:
# Calcular intervalos de 10 minutos
gabarito_tratado['inicio_intervalo'] = gabarito_tratado['datetime_partida'] - pd.Timedelta(seconds=600)
gabarito_tratado['fim_intervalo'] = gabarito_tratado['datetime_partida'] + pd.Timedelta(seconds=600)

In [85]:
# 1. Faça um left join usando 'id_veiculo' como chave
merged_df = gabarito_tratado.merge(solucao, on='id_veiculo', how='left', suffixes=('_gabarito', '_solucao'))

# 2. Filtrar as linhas com base nas condições de intervalo de tempo
resultado = merged_df[
    (merged_df['datetime_partida_solucao'] >= merged_df['inicio_intervalo']) & 
    (merged_df['datetime_partida_solucao'] <= merged_df['fim_intervalo'])
]

In [86]:
# 3. Remover linhas duplicadas
resultado = resultado.drop_duplicates()

# 4. Selecionar colunas específicas
resultado = resultado[['data', 'servico', 'sentido', 'id_veiculo', 'datetime_partida_gabarito', 
                       'datetime_chegada_gabarito', 'extensao', 'inicio_intervalo', 'fim_intervalo', 
                       'datetime_partida_solucao', 'datetime_chegada_solucao']]

resultado


data  servico sentido id_veiculo datetime_partida_gabarito  \
0      2023-01-02       10       I     A72037       2023-01-02 06:07:00   
259    2023-01-02       10       V     A72037       2023-01-02 06:29:00   
518    2023-01-02       10       I     A72037       2023-01-02 06:58:00   
777    2023-01-02       10       V     A72037       2023-01-02 07:28:00   
1036   2023-01-02       10       I     A72037       2023-01-02 08:02:00   
...           ...      ...     ...        ...                       ...   
208428 2023-01-03       10       I     A72151       2023-01-03 17:50:00   
208436 2023-01-03       10       V     A72151       2023-01-03 18:27:00   
208452 2023-01-03       10       V     A72151       2023-01-03 20:18:00   
208460 2023-01-03       10       I     A72151       2023-01-03 21:03:00   
208517 2023-01-03       10       I     A72085       2023-01-03 21:18:00   

       datetime_chegada_gabarito  extensao    inicio_intervalo  \
0            2023-01-02 06:22:00      5505 2023-01-02 05:57:00   
259          2023-01-02 06:56:00      4902 2023-01-02 06:19:00   
518          2023-01-02 07:16:00      5505 2023-01-02 06:48:00   
777          2023-01-02 07:58:00      4902 2023-01-02 07:18:00   
1036         2023-01-02 08:24:00      5505 2023-01-02 07:52:00   
...                          ...       ...                 ...   
208428       2023-01-03 18:25:00      5505 2023-01-03 17:40:00   
208436       2023-01-03 19:13:00      4902 2023-01-03 18:17:00   
208452       2023-01-03 21:00:00      4902 2023-01-03 20:08:00   
208460       2023-01-03 21:20:00      5505 2023-01-03 20:53:00   
208517       2023-01-03 21:37:00      5505 2023-01-03 21:08:00   

             fim_intervalo datetime_partida_solucao datetime_chegada_solucao  
0      2023-01-02 06:17:00      2023-01-02 06:07:57      2023-01-02 06:19:58  
259    2023-01-02 06:39:00      2023-01-02 06:32:00      2023-01-02 06:43:21  
518    2023-01-02 07:08:00      2023-01-02 07:00:04      2023-01-02 07:13:44  
777    2023-01-02 07:38:00      2023-01-02 07:31:27      2023-01-02 07:47:47  
1036   2023-01-02 08:12:00      2023-01-02 08:03:43      2023-01-02 08:22:04  
...                    ...                      ...                      ...  
208428 2023-01-03 18:00:00      2023-01-03 17:51:34      2023-01-03 18:22:35  
208436 2023-01-03 18:37:00      2023-01-03 18:30:36      2023-01-03 18:59:52  
208452 2023-01-03 20:28:00      2023-01-03 20:21:27      2023-01-03 20:45:28  
208460 2023-01-03 21:13:00      2023-01-03 21:04:37      2023-01-03 21:18:38  
208517 2023-01-03 21:28:00      2023-01-03 21:21:36      2023-01-03 21:35:16  

[699 rows x 11 columns]

Foram encontradas 699 das 706 viagens apresentadas no gabarito (94,75% das viagens do gabarito).

### 4. Identificar as viagens que não foram encontradas na solução e que estão no gabarito

In [105]:
tabela = gabarito_tratado.merge(resultado, how = 'left', 
                                left_on=['id_veiculo', 'datetime_partida'], 
                                right_on=['id_veiculo', 'datetime_partida_gabarito'])

linhas_nan = tabela[tabela['sentido'].isna()]
linhas_nan

id_veiculo    datetime_partida    datetime_chegada  inicio_intervalo_x  \
51      A72037 2023-01-04 06:00:00 2023-01-04 06:16:00 2023-01-04 05:50:00   
52      A72037 2023-01-04 06:28:00 2023-01-04 07:01:00 2023-01-04 06:18:00   
53      A72037 2023-01-04 07:03:00 2023-01-04 07:26:00 2023-01-04 06:53:00   
54      A72037 2023-01-04 07:29:00 2023-01-04 07:53:00 2023-01-04 07:19:00   
394     A72085 2023-01-04 06:30:00 2023-01-04 06:50:00 2023-01-04 06:20:00   
672     A72085 2023-01-15 07:20:00 2023-01-15 07:55:00 2023-01-15 07:10:00   
702     A72151 2023-01-03 19:19:00 2023-01-03 20:17:00 2023-01-03 19:09:00   

        fim_intervalo_x data  servico sentido datetime_partida_gabarito  \
51  2023-01-04 06:10:00  NaT      NaN     NaN                       NaT   
52  2023-01-04 06:38:00  NaT      NaN     NaN                       NaT   
53  2023-01-04 07:13:00  NaT      NaN     NaN                       NaT   
54  2023-01-04 07:39:00  NaT      NaN     NaN                       NaT   
394 2023-01-04 06:40:00  NaT      NaN     NaN                       NaT   
672 2023-01-15 07:30:00  NaT      NaN     NaN                       NaT   
702 2023-01-03 19:29:00  NaT      NaN     NaN                       NaT   

    datetime_chegada_gabarito  extensao inicio_intervalo_y fim_intervalo_y  \
51                        NaT       NaN                NaT             NaT   
52                        NaT       NaN                NaT             NaT   
53                        NaT       NaN                NaT             NaT   
54                        NaT       NaN                NaT             NaT   
394                       NaT       NaN                NaT             NaT   
672                       NaT       NaN                NaT             NaT   
702                       NaT       NaN                NaT             NaT   

    datetime_partida_solucao datetime_chegada_solucao  
51                       NaT                      NaT  
52                       NaT                      NaT  
53                       NaT                      NaT  
54                       NaT                      NaT  
394                      NaT                      NaT  
672                      NaT                      NaT  
702                      NaT                      NaT

Sete viagens que constam no gabarito não foram encontradas na solução e foram realizadas por três veículos diferentes. 

1) Quatro viagens do veículo A72037 no início do dia 04/01/2023

In [99]:
tabela['datetime_partida'] = pd.to_datetime(tabela['datetime_partida'])

tabela_A72037 = tabela[(tabela['id_veiculo'] == 'A72037') & (tabela['datetime_partida'].dt.date == pd.Timestamp('2023-01-04').date())]

tabela_A72037 = tabela_A72037[['id_veiculo', 'datetime_partida','datetime_chegada','datetime_partida_solucao','datetime_chegada_solucao']]
tabela_A72037

id_veiculo    datetime_partida    datetime_chegada  \
51     A72037 2023-01-04 06:00:00 2023-01-04 06:16:00   
52     A72037 2023-01-04 06:28:00 2023-01-04 07:01:00   
53     A72037 2023-01-04 07:03:00 2023-01-04 07:26:00   
54     A72037 2023-01-04 07:29:00 2023-01-04 07:53:00   
55     A72037 2023-01-04 08:01:00 2023-01-04 08:24:00   
56     A72037 2023-01-04 08:33:00 2023-01-04 09:03:00   
57     A72037 2023-01-04 09:07:00 2023-01-04 09:30:00   
58     A72037 2023-01-04 09:32:00 2023-01-04 10:01:00   
59     A72037 2023-01-04 10:05:00 2023-01-04 10:33:00   
60     A72037 2023-01-04 10:34:00 2023-01-04 11:04:00   
61     A72037 2023-01-04 11:08:00 2023-01-04 11:34:00   
62     A72037 2023-01-04 11:37:00 2023-01-04 11:58:00   
63     A72037 2023-01-04 12:02:00 2023-01-04 12:28:00   
64     A72037 2023-01-04 12:32:00 2023-01-04 12:53:00   
65     A72037 2023-01-04 12:55:00 2023-01-04 13:26:00   
66     A72037 2023-01-04 13:31:00 2023-01-04 14:07:00   
67     A72037 2023-01-04 14:09:00 2023-01-04 14:38:00   
68     A72037 2023-01-04 14:41:00 2023-01-04 15:12:00   
69     A72037 2023-01-04 15:14:00 2023-01-04 15:51:00   
70     A72037 2023-01-04 15:52:00 2023-01-04 16:25:00   
71     A72037 2023-01-04 16:27:00 2023-01-04 16:59:00   
72     A72037 2023-01-04 17:01:00 2023-01-04 17:36:00   
73     A72037 2023-01-04 17:39:00 2023-01-04 18:11:00   
74     A72037 2023-01-04 18:13:00 2023-01-04 18:49:00   
75     A72037 2023-01-04 18:49:00 2023-01-04 19:15:00   
76     A72037 2023-01-04 19:18:00 2023-01-04 19:45:00   
77     A72037 2023-01-04 19:47:00 2023-01-04 20:07:00   
78     A72037 2023-01-04 20:14:00 2023-01-04 20:41:00   
79     A72037 2023-01-04 20:42:00 2023-01-04 21:00:00   

   datetime_partida_solucao datetime_chegada_solucao  
51                      NaT                      NaT  
52                      NaT                      NaT  
53                      NaT                      NaT  
54                      NaT                      NaT  
55      2023-01-04 08:03:10      2023-01-04 08:22:11  
56      2023-01-04 08:38:58      2023-01-04 08:57:58  
57      2023-01-04 09:09:59      2023-01-04 09:28:19  
58      2023-01-04 09:38:20      2023-01-04 09:56:40  
59      2023-01-04 10:08:41      2023-01-04 10:28:42  
60      2023-01-04 10:40:02      2023-01-04 10:59:22  
61      2023-01-04 11:10:23      2023-01-04 11:32:24  
62      2023-01-04 11:41:24      2023-01-04 11:54:44  
63      2023-01-04 12:04:45      2023-01-04 12:26:25  
64      2023-01-04 12:35:33      2023-01-04 12:48:33  
65      2023-01-04 12:58:34      2023-01-04 13:22:54  
66      2023-01-04 13:34:09      2023-01-04 13:51:30  
67      2023-01-04 14:14:24      2023-01-04 14:36:44  
68      2023-01-04 14:44:25      2023-01-04 15:04:05  
69      2023-01-04 15:16:54      2023-01-04 15:46:35  
70      2023-01-04 15:56:36      2023-01-04 16:18:16  
71      2023-01-04 16:28:33      2023-01-04 16:56:53  
72      2023-01-04 17:04:54      2023-01-04 17:22:54  
73      2023-01-04 17:41:47      2023-01-04 18:07:08  
74      2023-01-04 18:16:28      2023-01-04 18:38:49  
75      2023-01-04 18:51:34      2023-01-04 19:11:35  
76      2023-01-04 19:20:15      2023-01-04 19:35:16  
77      2023-01-04 19:48:45      2023-01-04 20:05:06  
78      2023-01-04 20:17:58      2023-01-04 20:28:38  
79      2023-01-04 20:44:08      2023-01-04 20:58:48

Ao verificar a tabela de conformidade, também não foram encontradas as viagens antes de 08:00.

In [106]:
q = f"""
    SELECT
       *
     FROM
       `rj-smtr.projeto_subsidio_sppo.viagem_conformidade`
      WHERE
       data = "2023-01-04"
       AND id_veiculo = 'A72037'
"""

registros_conformidade_A72037 = bd.read_sql(q)

Downloading: 100%|██████████| 12/12 [00:00<00:00, 26.32rows/s]


In [110]:
registros_conformidade_A72037 = registros_conformidade_A72037[['id_veiculo','datetime_partida','datetime_chegada']]
registros_conformidade_A72037.sort_values(by='datetime_partida')

id_veiculo    datetime_partida    datetime_chegada
5      A72037 2023-01-04 08:39:58 2023-01-04 09:28:19
4      A72037 2023-01-04 09:39:40 2023-01-04 10:28:42
8      A72037 2023-01-04 10:40:42 2023-01-04 11:32:24
1      A72037 2023-01-04 11:41:44 2023-01-04 12:26:25
3      A72037 2023-01-04 12:35:53 2023-01-04 13:22:54
7      A72037 2023-01-04 13:34:49 2023-01-04 14:36:44
10     A72037 2023-01-04 14:44:45 2023-01-04 15:46:35
11     A72037 2023-01-04 15:56:36 2023-01-04 16:56:53
6      A72037 2023-01-04 17:05:34 2023-01-04 18:07:08
9      A72037 2023-01-04 18:16:48 2023-01-04 19:11:35
2      A72037 2023-01-04 19:20:35 2023-01-04 20:05:06
0      A72037 2023-01-04 20:18:18 2023-01-04 20:58:48

In [140]:
q = f"""
SELECT
  *
FROM
  `rj-smtr.br_rj_riodejaneiro_veiculos.gps_sppo`
WHERE
  DATA = "2023-01-04"
  AND id_veiculo = 'A72037'
  AND flag_em_operacao = true
"""

registros_gps_A72037 = bd.read_sql(q)

Downloading: 100%|██████████| 2577/2577 [00:00<00:00, 2975.75rows/s]


In [143]:
registros_gps_A72037 = registros_gps_A72037[['id_veiculo', 'hora','servico']]
registros_gps_A72037.sort_values(by='hora')

id_veiculo      hora servico
1654     A72037  06:47:25     007
2364     A72037  06:47:45     007
1355     A72037  06:48:06     007
1964     A72037  06:48:26     007
1403     A72037  06:48:46     007
...         ...       ...     ...
288      A72037  21:38:04     010
1547     A72037  21:38:24     010
1785     A72037  21:38:44     010
1341     A72037  21:39:04     010
683      A72037  21:39:24     010

[2577 rows x 3 columns]

Ao verificar a tabela com os sinais de GPS, entre 06:47 e 07:45, intervalo que é compatível com duas das quatro viagens não identificadas, foram identificados sinais do veículo operando no serviço 007, e não no serviço 010. Ou seja, das quatro viagens, duas não foram identificadas e as outras duas foram realizadas em outro serviço.

2) Duas viagens do veículo A72085: uma no dia 04/01/2023 e outra no dia 15/01/2023

2.1) Viagem do veículo A72085 no dia 04/01/2023

In [102]:
tabela_A72085_04_01 = tabela[(tabela['id_veiculo'] == 'A72085') & (tabela['datetime_partida'].dt.date == pd.Timestamp('2023-01-04').date())]

tabela_A72085_04_01 = tabela_A72085_04_01[['id_veiculo', 'datetime_partida','datetime_chegada','datetime_partida_solucao','datetime_chegada_solucao']]
tabela_A72085_04_01

id_veiculo    datetime_partida    datetime_chegada  \
394     A72085 2023-01-04 06:30:00 2023-01-04 06:50:00   
395     A72085 2023-01-04 06:59:00 2023-01-04 07:28:00   
396     A72085 2023-01-04 07:30:00 2023-01-04 07:50:00   
397     A72085 2023-01-04 07:59:00 2023-01-04 08:29:00   
398     A72085 2023-01-04 08:32:00 2023-01-04 09:00:00   
399     A72085 2023-01-04 09:03:00 2023-01-04 09:31:00   
400     A72085 2023-01-04 09:34:00 2023-01-04 10:04:00   
401     A72085 2023-01-04 10:07:00 2023-01-04 10:33:00   
402     A72085 2023-01-04 10:35:00 2023-01-04 11:04:00   
403     A72085 2023-01-04 11:06:00 2023-01-04 11:27:00   
404     A72085 2023-01-04 11:30:00 2023-01-04 11:58:00   
405     A72085 2023-01-04 12:03:00 2023-01-04 12:30:00   
406     A72085 2023-01-04 12:32:00 2023-01-04 13:03:00   
407     A72085 2023-01-04 13:05:00 2023-01-04 13:37:00   
408     A72085 2023-01-04 13:39:00 2023-01-04 14:05:00   
409     A72085 2023-01-04 14:06:00 2023-01-04 14:34:00   
410     A72085 2023-01-04 14:35:00 2023-01-04 15:04:00   
411     A72085 2023-01-04 15:06:00 2023-01-04 15:34:00   
412     A72085 2023-01-04 15:36:00 2023-01-04 16:08:00   
413     A72085 2023-01-04 16:10:00 2023-01-04 16:38:00   
414     A72085 2023-01-04 16:40:00 2023-01-04 17:07:00   
415     A72085 2023-01-04 17:16:00 2023-01-04 17:49:00   
416     A72085 2023-01-04 17:51:00 2023-01-04 18:15:00   
417     A72085 2023-01-04 18:22:00 2023-01-04 19:04:00   
418     A72085 2023-01-04 19:06:00 2023-01-04 19:28:00   
419     A72085 2023-01-04 19:38:00 2023-01-04 20:02:00   
420     A72085 2023-01-04 20:09:00 2023-01-04 20:29:00   
421     A72085 2023-01-04 20:32:00 2023-01-04 20:54:00   
422     A72085 2023-01-04 20:56:00 2023-01-04 21:14:00   

    datetime_partida_solucao datetime_chegada_solucao  
394                      NaT                      NaT  
395      2023-01-04 07:01:53      2023-01-04 07:14:14  
396      2023-01-04 07:31:57      2023-01-04 07:48:17  
397      2023-01-04 08:03:39      2023-01-04 08:22:40  
398      2023-01-04 08:34:00      2023-01-04 08:56:41  
399      2023-01-04 09:06:01      2023-01-04 09:22:02  
400      2023-01-04 09:37:39      2023-01-04 10:02:20  
401      2023-01-04 10:10:20      2023-01-04 10:28:01  
402      2023-01-04 10:38:01      2023-01-04 11:01:02  
403      2023-01-04 11:10:42      2023-01-04 11:25:03  
404      2023-01-04 11:32:03      2023-01-04 11:56:04  
405      2023-01-04 12:06:04      2023-01-04 12:25:24  
406      2023-01-04 12:34:45      2023-01-04 13:00:25  
407      2023-01-04 13:09:46      2023-01-04 13:28:06  
408      2023-01-04 13:42:47      2023-01-04 14:02:27  
409      2023-01-04 14:10:27      2023-01-04 14:27:08  
410      2023-01-04 14:38:29      2023-01-04 15:00:49  
411      2023-01-04 15:09:10      2023-01-04 15:25:30  
412      2023-01-04 15:40:21      2023-01-04 16:05:22  
413      2023-01-04 16:13:42      2023-01-04 16:32:23  
414      2023-01-04 16:43:39      2023-01-04 17:05:00  
415      2023-01-04 17:18:48      2023-01-04 17:35:28  
416      2023-01-04 17:57:09      2023-01-04 18:13:09  
417      2023-01-04 18:24:38      2023-01-04 18:39:39  
418      2023-01-04 19:07:38      2023-01-04 19:24:59  
419      2023-01-04 19:40:27      2023-01-04 19:53:07  
420      2023-01-04 20:11:04      2023-01-04 20:27:05  
421      2023-01-04 20:35:36      2023-01-04 20:47:17  
422      2023-01-04 20:57:35      2023-01-04 21:13:16

A viagem não foi encontrada na tabela de conformidade

In [121]:
q = f"""
    SELECT
       *
     FROM
       `rj-smtr.projeto_subsidio_sppo.viagem_conformidade`
      WHERE
       data = "2023-01-04"
       AND id_veiculo = 'A72085'
"""

registros_conformidade_A72085_04 = bd.read_sql(q)

Downloading: 100%|██████████| 14/14 [00:00<00:00, 39.55rows/s]


In [122]:
registros_conformidade_A72085_04 = registros_conformidade_A72085_04[['id_veiculo', 'datetime_partida','datetime_chegada']]
registros_conformidade_A72085_04.sort_values(by='datetime_partida')

id_veiculo    datetime_partida    datetime_chegada
2      A72085 2023-01-04 07:02:13 2023-01-04 07:48:17
8      A72085 2023-01-04 08:04:19 2023-01-04 08:56:41
11     A72085 2023-01-04 09:06:01 2023-01-04 10:02:20
6      A72085 2023-01-04 10:10:40 2023-01-04 11:01:02
5      A72085 2023-01-04 11:11:22 2023-01-04 11:56:04
9      A72085 2023-01-04 12:06:44 2023-01-04 13:00:25
7      A72085 2023-01-04 13:10:06 2023-01-04 14:02:27
10     A72085 2023-01-04 14:10:27 2023-01-04 15:00:49
13     A72085 2023-01-04 15:09:50 2023-01-04 16:05:22
12     A72085 2023-01-04 16:14:02 2023-01-04 17:05:00
4      A72085 2023-01-04 17:19:08 2023-01-04 18:13:09
3      A72085 2023-01-04 18:24:38 2023-01-04 19:24:59
0      A72085 2023-01-04 19:41:07 2023-01-04 20:27:05
1      A72085 2023-01-04 20:36:16 2023-01-04 21:13:16

In [137]:
q = f"""
SELECT
  *
FROM
  `rj-smtr.br_rj_riodejaneiro_veiculos.gps_sppo`
WHERE
  DATA = "2023-01-04"
  AND id_veiculo = 'A72085'
  AND flag_em_operacao = true
"""

registros_gps_A72085_04 = bd.read_sql(q)

Downloading: 100%|██████████| 2538/2538 [00:00<00:00, 2920.60rows/s]


In [139]:
registros_gps_A72085_04 = registros_gps_A72085_04[['id_veiculo', 'hora','servico']]
registros_gps_A72085_04.sort_values(by='hora')

id_veiculo      hora servico
2359     A72085  06:47:41     010
1745     A72085  06:48:01     010
1718     A72085  06:48:21     010
382      A72085  06:48:41     010
2155     A72085  06:49:01     010
...         ...       ...     ...
2294     A72085  21:59:27     010
1190     A72085  21:59:47     010
1070     A72085  22:00:07     010
125      A72085  22:00:27     010
955      A72085  22:00:47     010

[2538 rows x 3 columns]

Foi identificado sinal de GPS em operação para o veículo A72085 desde 06:47, horário próximo ao horário do final da viagem não identificada (06:50). Os sinais de GPS não contemplaram a duração total da viagem.

2.2) Viagem do veículo A72085 no dia 15/01/2023

In [127]:
tabela_A72085_04_15 = tabela[(tabela['id_veiculo'] == 'A72085') & (tabela['datetime_partida'].dt.date == pd.Timestamp('2023-01-15').date())]

tabela_A72085_04_15 = tabela_A72085_04_15[['id_veiculo', 'datetime_partida','datetime_chegada','datetime_partida_solucao','datetime_chegada_solucao']]
tabela_A72085_04_15

id_veiculo    datetime_partida    datetime_chegada  \
669     A72085 2023-01-15 06:36:00 2023-01-15 06:52:00   
670     A72085 2023-01-15 06:52:00 2023-01-15 07:04:00   
671     A72085 2023-01-15 07:05:00 2023-01-15 07:19:00   
672     A72085 2023-01-15 07:20:00 2023-01-15 07:55:00   
673     A72085 2023-01-15 08:01:00 2023-01-15 08:17:00   
674     A72085 2023-01-15 08:27:00 2023-01-15 09:02:00   
675     A72085 2023-01-15 09:06:00 2023-01-15 09:24:00   
676     A72085 2023-01-15 09:32:00 2023-01-15 09:52:00   
677     A72085 2023-01-15 10:02:00 2023-01-15 10:19:00   
678     A72085 2023-01-15 10:25:00 2023-01-15 10:49:00   
679     A72085 2023-01-15 10:58:00 2023-01-15 11:13:00   
680     A72085 2023-01-15 11:23:00 2023-01-15 11:46:00   
681     A72085 2023-01-15 11:52:00 2023-01-15 12:17:00   
682     A72085 2023-01-15 12:27:00 2023-01-15 12:48:00   
683     A72085 2023-01-15 12:51:00 2023-01-15 13:10:00   
684     A72085 2023-01-15 13:22:00 2023-01-15 13:55:00   
685     A72085 2023-01-15 13:56:00 2023-01-15 14:16:00   
686     A72085 2023-01-15 14:22:00 2023-01-15 14:48:00   
687     A72085 2023-01-15 14:49:00 2023-01-15 15:09:00   
688     A72085 2023-01-15 15:10:00 2023-01-15 15:40:00   
689     A72085 2023-01-15 15:42:00 2023-01-15 16:05:00   
690     A72085 2023-01-15 16:11:00 2023-01-15 16:38:00   
691     A72085 2023-01-15 16:40:00 2023-01-15 17:00:00   
692     A72085 2023-01-15 17:01:00 2023-01-15 17:43:00   
693     A72085 2023-01-15 17:44:00 2023-01-15 18:08:00   
694     A72085 2023-01-15 18:09:00 2023-01-15 18:49:00   
695     A72085 2023-01-15 18:51:00 2023-01-15 19:14:00   
696     A72085 2023-01-15 19:16:00 2023-01-15 19:38:00   
697     A72085 2023-01-15 19:41:00 2023-01-15 19:59:00   
698     A72085 2023-01-15 20:06:00 2023-01-15 20:28:00   
699     A72085 2023-01-15 20:30:00 2023-01-15 20:53:00   

    datetime_partida_solucao datetime_chegada_solucao  
669      2023-01-15 06:37:54      2023-01-15 06:50:34  
670      2023-01-15 06:54:54      2023-01-15 07:03:34  
671      2023-01-15 07:06:55      2023-01-15 07:17:15  
672                      NaT                      NaT  
673      2023-01-15 08:03:32      2023-01-15 08:15:12  
674      2023-01-15 08:31:13      2023-01-15 08:43:53  
675      2023-01-15 09:08:37      2023-01-15 09:20:17  
676      2023-01-15 09:35:39      2023-01-15 09:46:39  
677      2023-01-15 10:04:37      2023-01-15 10:16:37  
678      2023-01-15 10:30:31      2023-01-15 10:44:51  
679      2023-01-15 11:00:27      2023-01-15 11:10:27  
680      2023-01-15 11:26:51      2023-01-15 11:41:32  
681      2023-01-15 11:55:16      2023-01-15 12:11:37  
682      2023-01-15 12:32:07      2023-01-15 12:43:27  
683      2023-01-15 12:54:08      2023-01-15 13:07:48  
684      2023-01-15 13:27:52      2023-01-15 13:43:32  
685      2023-01-15 13:58:36      2023-01-15 14:13:17  
686      2023-01-15 14:25:57      2023-01-15 14:40:58  
687      2023-01-15 14:50:59      2023-01-15 15:06:59  
688      2023-01-15 15:18:20      2023-01-15 15:33:00  
689      2023-01-15 15:43:41      2023-01-15 16:01:32  
690      2023-01-15 16:14:13      2023-01-15 16:31:33  
691      2023-01-15 16:42:14      2023-01-15 16:57:34  
692      2023-01-15 17:05:54      2023-01-15 17:33:20  
693      2023-01-15 17:46:33      2023-01-15 18:06:13  
694      2023-01-15 18:13:54      2023-01-15 18:35:34  
695      2023-01-15 18:52:02      2023-01-15 19:11:03  
696      2023-01-15 19:18:23      2023-01-15 19:32:43  
697      2023-01-15 19:42:45      2023-01-15 19:57:06  
698      2023-01-15 20:08:57      2023-01-15 20:20:57  
699      2023-01-15 20:32:29      2023-01-15 20:50:30

In [131]:
q = f"""
    SELECT
       *
     FROM
       `rj-smtr.projeto_subsidio_sppo.viagem_conformidade`
      WHERE
       data = "2023-01-15"
       AND id_veiculo = 'A72085'
"""

registros_conformidade_A72085_15 = bd.read_sql(q)

Downloading: 0rows [00:00, ?rows/s]


In [132]:
registros_conformidade_A72085_15 

Empty DataFrame
Columns: [id_viagem, data, id_empresa, id_veiculo, servico_informado, servico_realizado, distancia_planejada, sentido, datetime_partida, datetime_chegada, trip_id, shape_id, tempo_viagem, distancia_aferida, distancia_inicio_fim, n_registros_middle, n_registros_start, n_registros_end, n_registros_out, n_registros_total, n_registros_minuto, n_registros_shape, perc_conformidade_shape, perc_conformidade_distancia, perc_conformidade_registros, versao_modelo]
Index: []

[0 rows x 26 columns]

Não foram encontradas viagens para o veículo A72085 no dia 15-01-2023 na tabela de viagem_conformidade. 

In [148]:
q = f"""
SELECT
  *
FROM
  `rj-smtr.br_rj_riodejaneiro_veiculos.gps_sppo`
WHERE
  DATA = "2023-01-15"
  AND id_veiculo = 'A72085'
  AND flag_em_operacao = true
"""

registros_gps_A72085_15 = bd.read_sql(q)

Downloading: 100%|██████████| 2455/2455 [00:00<00:00, 2746.05rows/s]


In [149]:
registros_gps_A72085_15 = registros_gps_A72085_15[['id_veiculo', 'hora','servico']]

hora_inicial = datetime.strptime('07:20', '%H:%M').time()
hora_final = datetime.strptime('07:55', '%H:%M').time()

registros_gps_A72085_15 = registros_gps_A72085_15[(registros_gps_A72085_15['hora'] >= hora_inicial) & (registros_gps_A72085_15['hora'] <= hora_final)]

registros_gps_A72085_15.sort_values(by='hora')

id_veiculo      hora servico
657      A72085  07:20:03     010
1782     A72085  07:20:23     010
2154     A72085  07:20:43     010
419      A72085  07:21:03     010
369      A72085  07:21:23     010
...         ...       ...     ...
259      A72085  07:53:22     010
1001     A72085  07:53:42     010
520      A72085  07:54:02     010
864      A72085  07:54:22     010
1222     A72085  07:54:42     010

[72 rows x 3 columns]

Foram encontrados sinais de GPS para o mesmo veículo, dia e serviço.

3) Uma viagem do veículo A72151 no dia 03/01/2023

In [150]:
tabela_A72151 = tabela[(tabela['id_veiculo'] == 'A72151') & (tabela['datetime_partida'].dt.date == pd.Timestamp('2023-01-03').date())]
tabela_A72151 = tabela_A72151[['id_veiculo', 'datetime_partida','datetime_chegada','datetime_partida_solucao','datetime_chegada_solucao']]
tabela_A72151

id_veiculo    datetime_partida    datetime_chegada  \
700     A72151 2023-01-03 17:50:00 2023-01-03 18:25:00   
701     A72151 2023-01-03 18:27:00 2023-01-03 19:13:00   
702     A72151 2023-01-03 19:19:00 2023-01-03 20:17:00   
703     A72151 2023-01-03 20:18:00 2023-01-03 21:00:00   
704     A72151 2023-01-03 21:03:00 2023-01-03 21:20:00   

    datetime_partida_solucao datetime_chegada_solucao  
700      2023-01-03 17:51:34      2023-01-03 18:22:35  
701      2023-01-03 18:30:36      2023-01-03 18:59:52  
702                      NaT                      NaT  
703      2023-01-03 20:21:27      2023-01-03 20:45:28  
704      2023-01-03 21:04:37      2023-01-03 21:18:38

Checar a tabela de conformidade

In [151]:
q = f"""
    SELECT
       *
     FROM
       `rj-smtr.projeto_subsidio_sppo.viagem_conformidade`
      WHERE
       data = "2023-01-03"
       AND id_veiculo = 'A72151'
"""

registros_conformidade_A72151 = bd.read_sql(q)

Downloading: 100%|██████████| 17/17 [00:00<00:00, 47.75rows/s]


In [153]:
registros_conformidade_A72151.sort_values(by = 'datetime_partida')

id_viagem        data id_empresa id_veiculo  \
0   A72151-006-I-fm9l-20230103060937  2023-01-03        721     A72151   
3   A72151-006-V-wu1v-20230103063636  2023-01-03        721     A72151   
9   A72151-006-I-fm9l-20230103071644  2023-01-03        721     A72151   
10  A72151-006-V-wu1v-20230103075032  2023-01-03        721     A72151   
1   A72151-006-I-fm9l-20230103082749  2023-01-03        721     A72151   
15  A72151-006-V-wu1v-20230103085921  2023-01-03        721     A72151   
5   A72151-006-I-fm9l-20230103094518  2023-01-03        721     A72151   
14  A72151-006-V-wu1v-20230103101638  2023-01-03        721     A72151   
6   A72151-006-I-fm9l-20230103110541  2023-01-03        721     A72151   
2   A72151-006-V-wu1v-20230103114012  2023-01-03        721     A72151   
7   A72151-006-I-fm9l-20230103122028  2023-01-03        721     A72151   
13  A72151-006-V-wu1v-20230103125502  2023-01-03        721     A72151   
12  A72151-007-I-e2fc-20230103151414  2023-01-03        721     A72151   
11  A72151-007-V-tdv7-20230103155356  2023-01-03        721     A72151   
8   A72151-007-I-e2fc-20230103170732  2023-01-03        721     A72151   
16  A72151-010-C-01o6-20230103183121  2023-01-03        721     A72151   
4   A72151-010-C-01o6-20230103202227  2023-01-03        721     A72151   

   servico_informado servico_realizado  distancia_planejada sentido  \
0                006               006                8.427       I   
3                006               006                8.885       V   
9                006               006                8.427       I   
10               006               006                8.885       V   
1                006               006                8.427       I   
15               006               006                8.885       V   
5                006               006                8.427       I   
14               006               006                8.885       V   
6                006               006                8.427       I   
2                006               006                8.885       V   
7                006               006                8.427       I   
13               006               006                8.885       V   
12               007               007                8.414       I   
11               007               007                9.196       V   
8                007               007                8.414       I   
16               010               010               10.406       C   
4                010               010               10.406       C   

      datetime_partida    datetime_chegada  ... n_registros_start  \
0  2023-01-03 06:09:37 2023-01-03 06:30:36  ...                 1   
3  2023-01-03 06:36:36 2023-01-03 06:59:37  ...                 1   
9  2023-01-03 07:16:44 2023-01-03 07:43:17  ...                 1   
10 2023-01-03 07:50:32 2023-01-03 08:17:38  ...                 1   
1  2023-01-03 08:27:49 2023-01-03 08:50:50  ...                 1   
15 2023-01-03 08:59:21 2023-01-03 09:30:06  ...                 1   
5  2023-01-03 09:45:18 2023-01-03 10:09:08  ...                 1   
14 2023-01-03 10:16:38 2023-01-03 10:47:10  ...                 1   
6  2023-01-03 11:05:41 2023-01-03 11:30:27  ...                 1   
2  2023-01-03 11:40:12 2023-01-03 12:03:13  ...                 1   
7  2023-01-03 12:20:28 2023-01-03 12:45:31  ...                 1   
13 2023-01-03 12:55:02 2023-01-03 13:23:34  ...                 1   
12 2023-01-03 15:14:14 2023-01-03 15:42:50  ...                 1   
11 2023-01-03 15:53:56 2023-01-03 16:21:53  ...                 1   
8  2023-01-03 17:07:32 2023-01-03 17:33:18  ...                 1   
16 2023-01-03 18:31:21 2023-01-03 20:15:42  ...                 2   
4  2023-01-03 20:22:27 2023-01-03 21:18:38  ...                 2   

   n_registros_end  n_registros_out  n_registros_total  n_registros_minuto  \
0                1                0                 84                  22   
3               

Foi encontrada uma viagem na linha 010 entre 18:31 e 20:15, horário que se sobrepõe ao horário da viagem apontada no gabarito, entre 19:19 e 20:17.

5. Viagens que foram encontradas na solução, mas que não foram encontradas no gabarito.

Trinta e cinco viagens que constam na solução não foram encontradas no gabarito.

In [154]:
tabela = solucao.merge(resultado, how = 'left', 
                                left_on=['id_veiculo', 'datetime_partida'], 
                                right_on=['id_veiculo', 'datetime_partida_solucao'])

tabela

linhas_nan = tabela[tabela['sentido_y'].isna()]
linhas_nan
len(linhas_nan)

35

### 6. Conclusão

Ao considerar um intervalo de 10 minutos para mais ou para menos entre os horários de partida, foram encontradas 699 das 706 viagens apresentadas no gabarito (94,75% das viagens do gabarito).

Ao analisar as sete viagens não identificadas na solução acima, concluiu-se que:

- Não foram encontrados dados das quatro viagens do veículo A72037 na linha 010, mas foram encontrados sinais de GPS em horários compatíveis com duas das quatro viagens, operando na linha 007.

- No caso do veículo A72085, na viagem do dia 15/01/2023 foram encontrados sinais de GPS para o mesmo veículo, dia e serviço (mas não foram encontrados dados para este dia na tabela de conformidade). Quanto a viagem do dia 04/01/2023, foram identificados sinais de GPS apenas em horário próximo ao término da viagem.

- Para o veículo A72151, foi encontrada uma viagem da linha 010 entre 18:31 e 20:15, horário que se sobrepõe ao horário da viagem apontada no gabarito, entre 19:19 e 20:17.